pkg/stream/connpool/msgconnpool/connpool.go

```diff
@@ -0,0 +1,347 @@
+}
+
+// Destroy the pool
+func (p *connpool) Destroy() {
+	atomic.StoreUint64(&p.destroyed, 1)
+	p.client.host.Connection.Close(api.NoFlush, api.LocalClose)
+}
+
+// types.StreamEventListener
+// types.ConnectionEventListener
+// types.StreamConnectionEventListener
+type activeClient struct {
+	pool *connpool
+	host *types.CreateConnectionData
+
+	keepAlive *keepAliveListener
+
+	reconnectLock sync.Mutex
+
+	reconnectBackoff    []time.Duration
+	connectTryTimes     int
+}
+
+func (ac *activeClient) OnEvent(event api.ConnectionEvent) {
+	//  all close event:
+	//  ce == LocalClose || ce == RemoteClose ||
+	//	ce == OnReadErrClose || ce == OnWriteErrClose || ce == OnWriteTimeout
+	switch event {
+	case api.RemoteClose:
+		goto RECONN
+	case api.OnReadErrClose, api.OnWriteErrClose, api.OnWriteTimeout,
+		api.ConnectTimeout, api.ConnectFailed, api.LocalClose:
+		// RemoteClose when read/write error
+		// LocalClose when there is a panic
+		// OnReadErrClose when read failed
+		goto RECONN
+
+	default:
+		return
+	}
+
+RECONN:
+	if ac.pool.autoReconnectWhenClose {
+		// auto reconnect when close
+		log.DefaultLogger.Warnf("[connpool] reconnect after event : %v,  host : %v", event, ac)
+		ac.reconnect()
+	} else {
+		ac.removeFromPool()
+	}
+}
+
+// generate the client, and set it to the connpool
+func (p *connpool) initActiveClient() {
+	p.client = &activeClient{
+		pool: p,
+		reconnectBackoff: []time.Duration{
+			0, time.Second,
+			time.Second * 2,
+			time.Second * 5,
+			time.Second * 10,
+		},
+	}
+
+	p.client.initConnection()
+}
+
+func (ac *activeClient) initConnection() {
+	// build new conn
+	// must create this new conn, the same conn can only be connected once
+	createConnData := ac.pool.Host().CreateConnection(context.Background())
+
+	// event listener start
+	// createConnData.Connection.AddConnectionEventListener(ac)
+	createConnData.Connection.AddConnectionEventListener(ac)
+	// event listener end
+
+	// connect the new connection
+	err := createConnData.Connection.Connect()
+
+	if err != nil {
+		if ac.host == nil { // the first time
+			// atomic store, avoid partial write
+			atomic.StorePointer((*unsafe.Pointer)(unsafe.Pointer(&ac.host)), unsafe.Pointer(&createConnData))
+			log.DefaultLogger.Warnf("[connpool] connect failed %v times, host : %p",
+				ac.connectTryTimes, ac.host.Host.AddressString())
+		} else {
+			log.DefaultLogger.Warnf("[connpool] reconnect failed %v times, ac : %p",
+				ac.connectTryTimes, ac)
+		}
+
+		return
+	}
+
+	// atomic store, avoid partial write
+	atomic.StorePointer((*unsafe.Pointer)(unsafe.Pointer(&ac.host)), unsafe.Pointer(&createConnData))
+
+	log.DefaultLogger.Infof("[connpool] reconnect succeed after %v tries, host %v, ac %p, host: %v",
+		ac.connectTryTimes, ac.host.Host.AddressString(), ac, ac.host)
+
+	// if pool was destroyed, but connection was connected
+	// we need to close it
+	if atomic.LoadUint64(&ac.pool.destroyed) == 1 {
+		ac.host.Connection.Close(api.NoFlush, api.LocalClose)
```

Comment:
`ac.pool.destroyed == 1` means that has been called `Destroy()` function，why called Close again?

Reply:
reconnect -> Destroy -> reconnect success && store pointer to ac.host
then a connection will be leaked